subprojects/core/src/main/java/org/gradle/internal/cleanup/DefaultBuildOutputCleanupCache.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.cleanup;
+
+import org.gradle.cache.CacheBuilder;
+import org.gradle.cache.CacheRepository;
+import org.gradle.cache.PersistentCache;
+import org.gradle.cache.internal.FileLockManager;
+import org.gradle.internal.IoActions;
+import org.gradle.util.GFileUtils;
+import org.gradle.util.GradleVersion;
+
+import java.io.File;
+import java.util.Collections;
+
+import static org.gradle.cache.internal.filelock.LockOptionsBuilder.mode;
+
+public class DefaultBuildOutputCleanupCache implements BuildOutputCleanupCache {
+
+    private final CacheRepository cacheRepository;
+    private final File cacheDir;
+    private final BuildOutputDeleter buildOutputDeleter;
+    private final BuildOutputCleanupRegistry buildOutputCleanupRegistry;
+
+    public DefaultBuildOutputCleanupCache(CacheRepository cacheRepository, File cacheBaseDir, BuildOutputDeleter buildOutputDeleter, BuildOutputCleanupRegistry buildOutputCleanupRegistry) {
+        this.cacheRepository = cacheRepository;
+        this.cacheDir = new File(cacheBaseDir, ".gradle/noVersion/buildOutputCleanup");
```

Comment:
I think this isn't quite right. This is supposed to be the project cache dir? The project cache dir can be something other than .gradle/, I think we're supposed to use a different CacheRepository method to make this cache Gradle or Project scope.